text/0000-env-root-not-optional.md

```diff
@@ -0,0 +1,96 @@
+kind of hard to justify.
+
+```pony
+
+actor Main
+  new create(env: Env) =>
+    try
+      let auth: AmbientAuth = env.root as AmbientAuth
+      // do something with auth
+    else
+      env.err.print("no auth available.")
+      env.exitcode(1)
+    end
+```
+
+```pony
+actor Main
+  new create(env: Env) =>
+    match env.root
+    | let auth: AmbientAuth =>
+      // do something with auth
+    else
+      env.err.print("no auth available.")
+      env.exitcode(1)
+    end
+```
+
+[`Env.root`](https://stdlib.ponylang.org/builtin-Env#let-root-ambientauth-val-none-val) is of type `(AmbientAuth | None)`
+and can thus possibly be `None`. This is for enabling people to create artificial `Env` instances that not need an `AmbientAuth` e.g. for testing
+handling of stdout via `Env.out` or similar things.
+
+But after searching for usages of artificial `Env` instances, it turned out, this feature is actually not used at all.
+There might be several reasons for this:
+
+ * Pieces of code that use command line arguments, environment variables, out-streams or other members of `Env` take those in their constructor or methods
+   and artificial instances of those can be easily used for testing without an artificial `Env`.
+ * There might be no need to use an artificial env without `AmbientAuth` as we have `Env` available in ponytest
+   anyways via [`TestHelper.env`](https://stdlib.ponylang.org/ponytest-TestHelper#let-env-env-val), including `AmbientAuth`.
+
+In the light of this, it seems more reasonable to optimize for convenience of usage in the normal case than
+designing for possible use cases, that are actually not used at all, at the price of added ceremony/boilerplate.
+
+It might be desired to control what kind of control a pony process is started with (maybe it is not allowed to write to the filesystem) or
+to actually give no auth at all to the process via some cli flag and some runtime code, determining what kind of auth to put into `env.root`.
+But this change requires a more substantial change to both `Env` and the runtime, so this might be left for some future version of pony
+and this broader idea should not interfere with what this RFC is suggesting.
+
+# Detailed design
+
+Make [`Env.root`](https://stdlib.ponylang.org/builtin-Env#let-root-ambientauth-val-none-val) have type `AmbientAuth`
+and change [`Env.create`](https://stdlib.ponylang.org/builtin-Env/#create) to take an `AmbientAuth` for the `root'` parameter,
+thus require `AmbientAuth` when creating an artifical instance of `Env`.
+
+# How We Teach This
+
+Next to substantial changes in stdlib documentation and tutorial, examples and patterns, adapting the examples to how `Env.root` can be accessed now, 
+a small change in the docstrings of `Env`, correcting how to create an `Env` should be enough.
+
+# How We Test This
+
+# Drawbacks
+
+* Breaks existing code
+* Makes creating artificial `Env` instances a little harder.
```

Comment:
How would you create an artificial env without this? "a little harder" is very vague.

Reply:
How _would_ you create eg. a testing `Env` that doesn't need `AmbientAuth` under this proposal? As far as I can tell that'd be impossible, and I'm not exacty convinced we'd even want every test `Env` to have `AmbientAuth` when they don't need it

Reply:
An `AmbienAuth` would be required to create an artifical `Env` with this RFC. The `root` argument to `Env.create` will be changed from `(AmbienthAuth | None)` to `AmbientAuth`.

Currently in every ponytest, one has the process `Env` available as `TestHelper.env`. From the current layout of things, one can create artifical `Env` instances in tests.

If the current ponytest design should be changed, this is a completely different discussion. Which i am totally happy to have (in another context though).